cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,1164 @@
+                 edges,
+                 edges + num_invalid_edges,
+                 get_dataframe_buffer_begin(invalid_edges_buffer));
+    
+    // sort back the edges as those are needed later when running a binary tree
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        edges,
+                        edges + num_edges,
+                        num_triangles.begin());
+
+
+    // Unroll and remove/mask edges as long as there are still edges part
+    // of the K-Truss.
+    auto num_valid_edges = num_edges - num_invalid_edges;
+    while ((num_valid_edges != 0) && (num_invalid_edges !=0)) {
+      // case 2: unroll (q, r)
+      // FIXME: Update the num_edges when removing edges
+
+      
+      // FIXME: Need a buffer for the incomming vertex pairs because 'edges_to_num_triangles'
+      // is sorted in a way that matches the number of triangles per edges therefore,
+      // can't use a zip iterator for 'incoming_vertex_pairs'. This adds increase memory
+      // footprint
+      auto incoming_vertex_pairs = allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(
+        num_edges, handle.get_stream());
+
+      thrust::tabulate(handle.get_thrust_policy(),
+                       get_dataframe_buffer_begin(incoming_vertex_pairs),
+                       get_dataframe_buffer_end(incoming_vertex_pairs),
+                       [edges=edges
+                       ]
+                       __device__(edge_t idx){
+                          auto edge = thrust::make_tuple(thrust::get<1>(*(edges + idx)), thrust::get<0>(*(edges + idx)));
+                          return edge;
+                          
+                       });
+      
+      // Sort the 'incoming_vertex_pairs' by 'dst'
+      thrust::sort(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(incoming_vertex_pairs),
+        get_dataframe_buffer_end(incoming_vertex_pairs));
+
+      // For each (q, r) edges to unroll, find the incoming edges to 'r' let's say from 'p' and
+      // create the pair (p, q)
+      rmm::device_uvector<int> prefix_sum(num_invalid_edges + 1, handle.get_stream());
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        prefix_sum.begin(),
+        prefix_sum.end(),
+        [invalid_first   = get_dataframe_buffer_begin(invalid_edges_buffer),
+         dst_array_begin = std::get<0>(incoming_vertex_pairs).begin(),
+         num_edges       = num_edges] __device__(auto idx) {
+          auto src           = thrust::get<0>(*(invalid_first + idx));
+          auto dst           = thrust::get<1>(*(invalid_first + idx));
+          auto dst_array_end = dst_array_begin + num_edges;
+          auto itr_lower = thrust::lower_bound(thrust::seq, dst_array_begin, dst_array_end, dst);
+          auto idx_lower =
+            thrust::distance(dst_array_begin, itr_lower);  // FIXME: remove self loops
+          auto itr_upper = thrust::upper_bound(thrust::seq, dst_array_begin, dst_array_end, dst);
```

Comment:
thrust::upper_bound(thrust::seq, itr_lower, dst_array_end, dst) to limit search range.